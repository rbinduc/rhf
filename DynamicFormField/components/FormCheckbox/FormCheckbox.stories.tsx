import React from 'react';
import {storiesOf} from '@storybook/react';
import FormCheckbox, {IFormCheckboxProps} from '.';
import useFormHook from '../../utils/hooks/useFormHook';
import {themeMock} from '../../../../../themes/theme-mock';
import {FormProvider} from 'react-hook-form';
import {ThemeProvider} from 'emotion-theming';

const defaultProps: IFormCheckboxProps = {
  id: 'formCheckbox',
  name: 'formCheckbox',
  label: 'Form Checkbox',
};

storiesOf('Organisms/RHF/DynamicForm/FormCheckbox', module)
  .addDecorator((getStory) => <ThemeProvider theme={themeMock}>{getStory()}</ThemeProvider>)
  .add('default case', () => {
    const {methods}: any = useFormHook();
    return (
      <FormProvider {...methods}>
        <FormCheckbox {...defaultProps} />
      </FormProvider>
    );
  });
