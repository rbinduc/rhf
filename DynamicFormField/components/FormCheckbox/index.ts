export {default} from './FormCheckbox';
export type {IFormCheckboxProps} from './FormCheckbox';
