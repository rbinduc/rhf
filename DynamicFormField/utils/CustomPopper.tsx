import React from 'react';
import {Popper, PopperProps} from '@material-ui/core';

export const CustomPopper = ({placement = 'bottom', ...rest}: PopperProps) => <Popper {...rest} placement={placement} />;
