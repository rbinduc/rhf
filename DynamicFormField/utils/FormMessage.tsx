import styled from '@emotion/styled';
import {FormHelperText} from '@material-ui/core';
import React from 'react';
import {IconNames} from '../../../../icons/IconNameConstants';
import {TenantVector} from '../../../../icons/TenantVector/TenantVector.component';

interface IMessageProps {
  message?: React.ReactNode;
  isBold?: boolean;
  isLarge?: boolean;
}

export interface IFormMessageProps extends IMessageProps {
  id?: string;
  Icon?: IconNames;
  component?: React.ElementType;
  margin?: 'dense';
  isError?: boolean;
  hideIcon?: boolean;
}

const FormMessage = ({id = '', Icon, message, component, margin, isBold, isLarge, hideIcon, isError = true}: IFormMessageProps) =>
  message ? (
    <FormHelperText id={id} error={!!isError} component={component || 'div'} margin={margin}>
      <FormMessageWrapper>
        {!hideIcon && (
          <span id={id + '_icon'}>
            <TenantVector tenantVectorName={Icon || (isError ? IconNames.FieldError : IconNames.Info)} />
          </span>
        )}
        <Message id={id + '_message'} isBold={!!isBold} isLarge={!!isLarge}>
          {message}
        </Message>
      </FormMessageWrapper>
    </FormHelperText>
  ) : null;

const FormMessageWrapper = styled.span`
  display: flex;
  flex-direction: row;
  & svg {
    margin-top: 0.2rem;
    margin-right: 0.5rem;
  }
  word-break: break-word;
`;

const Message = styled.span<IMessageProps>`
  font-weight: ${({isBold}) => (isBold ? 'bold' : 400)};
  font-size: ${({isLarge}) => (isLarge ? '1rem' : '0.875rem')};
`;

export default FormMessage;
