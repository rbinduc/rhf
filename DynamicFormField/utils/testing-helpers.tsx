import React from 'react';
import {render} from '@testing-library/react';
import {FormProvider} from 'react-hook-form';
import useFormHook from './hooks/useFormHook';
import {ThemeProvider} from 'emotion-theming';
import {themeMock} from '../../../../themes/theme-mock';

interface IDefaultValue {
  [key: string]: any;
}

interface RenderOptions {
  defaultValues?: IDefaultValue;
  toPassBack?: any[];
}

/**
 * Testing Library utility function to wrap tested component in React Hook Form
 * @param {ReactElement} ui A React component
 * @param objectParameters
 * @param {Object} objectParameters.defaultValues Initial form values to pass into
 * React Hook Form, which you can then assert against
 * @param {Array} objectParameters.toPassBack React Hook Form method names which we'd
 * like to pass back and use in tests.  A primary use case is sending back 'setError',
 * so we can manually setErrors on React Hook Form components and test error handling
 */
export function renderWithDyanmicForm(
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
  {defaultValues = {}, toPassBack = []}: RenderOptions = {}
) {
  let dynamicFormMethods = {} as any;

  const Wrapper = ({children}: {children: React.ReactNode}) => {
    const {methods}: any = useFormHook({defaultValues});
    for (let dynamicFormItem of toPassBack) {
      dynamicFormMethods[dynamicFormItem] = methods[dynamicFormItem];
    }

    return (
      <ThemeProvider theme={themeMock}>
        <FormProvider {...methods}>{children}</FormProvider>
      </ThemeProvider>
    );
  };

  return {
    ...render(ui, {wrapper: Wrapper as React.ComponentType<{}>}),
    dynamicFormMethods,
  };
}

/**
 * Higher order helper function which wraps a component w/ React Hook Form
 * @param {React Component} WrappedComponent to pass into
 * @param {*} restProps any other remaining props
 * @returns {React Component}
 */
export function withDynamicForm(WrappedComponent: React.ComponentType<any>, restProps: object, defaultValues: IDefaultValue) {
  return () => {
    const methods = useFormHook({defaultValues});

    return (
      <ThemeProvider theme={themeMock}>
        <WrappedComponent {...restProps} {...methods} />
      </ThemeProvider>
    );
  };
}
