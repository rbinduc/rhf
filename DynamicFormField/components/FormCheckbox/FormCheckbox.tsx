import React from 'react';
import Checkbox, {CheckboxProps} from '@material-ui/core/Checkbox';
import {FormControl, FormControlLabel, FormLabel} from '@material-ui/core';
import {Controller, useFormContext} from 'react-hook-form';
import {getLabel, getRequiredLabel, IDynamicFormFieldMode, IValidations} from '../../DynamicFormField';
import FormMessage from '../../utils/FormMessage';
import {CheckedIcon} from '../../../../../icons/global/checkbox/Checked';
import {UncheckedIcon} from '../../../../../icons/global/checkbox/Unchecked';
import ReactHtmlParser from 'react-html-parser';
import styled from '@emotion/styled';

// ----------------------------------------------------------------------------
// Props
// ----------------------------------------------------------------------------
export interface IFormCheckboxProps extends CheckboxProps {
  label: React.ReactNode;
  validations?: IValidations;
  isDisabled?: boolean;
  isRequired?: boolean;
  mode?: IDynamicFormFieldMode;
  labelId?: string;
  hideAsterisk?: boolean;
}

// ----------------------------------------------------------------------------
// Component
// ----------------------------------------------------------------------------
const FormCheckbox: React.FC<IFormCheckboxProps> = ({
  name,
  label,
  validations,
  isDisabled,
  defaultValue,
  isRequired = false,
  title,
  mode,
  labelId = '',
  hideAsterisk = false,
  ...other
}: IFormCheckboxProps) => {
  const {
    control,
    formState: {errors},
  } = useFormContext();

  /**
   * We need custom logic to check if checkbox is required since isRequired is not enough,
   * so we need to check if value is actually set to boolean value 'true'.
   */
  const rules = {
    ...validations,
    validate: (value: boolean) => (isRequired ? value : true),
  };

  const errorMessage = errors[name!]?.message;
  return (
    <FormControlStyled error={Boolean(errorMessage)} required={isRequired} shrink={mode === IDynamicFormFieldMode.DENSE}>
      {title && <FormLabelStyled component="legend">{title}</FormLabelStyled>}
      <Controller
        render={({field}) => (
          <FormControlLabelStyled
            control={
              <CheckboxStyled
                {...other}
                icon={<UncheckedIcon />}
                checkedIcon={<CheckedIcon />}
                disabled={isDisabled}
                inputProps={{
                  'aria-label': label && typeof label === 'string' ? label : '',
                  'aria-hidden': isDisabled,
                  ...other.inputProps,
                }}
                color={'primary'}
              />
            }
            label={
              typeof label === 'string' ? (
                ReactHtmlParser(!!title ? label : getLabel(label, isRequired, hideAsterisk))
              ) : (
                <>
                  {label}
                  {getRequiredLabel(isRequired, hideAsterisk)}
                </>
              )
            }
            onChange={(e) => field.onChange((e.target as HTMLInputElement).checked)}
            checked={field.value}
            id={labelId}
          />
        )}
        name={name!}
        control={control}
        rules={rules}
        defaultValue={defaultValue}
      />
      <FormMessage message={errorMessage} margin="dense" />
    </FormControlStyled>
  );
};

interface IFormControlStyled {
  shrink?: boolean;
}

const FormControlStyled = styled(({shrink, ...rest}) => <FormControl {...rest} />)<IFormControlStyled>`
  & .MuiFormControlLabel-label {
    line-height: 1.3;
    font-size: ${(props) => (props.shrink ? '0.875rem' : '1rem')};
  }
`;

const FormLabelStyled = styled(FormLabel)<any>`
  font-size: 1rem;
  font-style: normal;
  font-weight: 700;
  line-height: 1.375rem;
  letter-spacing: 0;
  text-align: left;
  padding-bottom: 0.625rem;
`;

const FormControlLabelStyled = styled(FormControlLabel)`
  align-items: flex-start;
`;

const CheckboxStyled = styled(Checkbox)`
  &.MuiButtonBase-root {
    padding-top: 0;
    padding-bottom: 0;
  }
  &.MuiFormControlLabel-root {
    align-items: flex-start;
  }
`;

export default FormCheckbox;
