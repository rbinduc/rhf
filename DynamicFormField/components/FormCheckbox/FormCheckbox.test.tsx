import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import {screen, act} from '@testing-library/react';
import FormCheckbox, {IFormCheckboxProps} from './FormCheckbox';
import {renderWithDyanmicForm} from '../../utils/testing-helpers';
import {UseFormReturn} from 'react-hook-form';

const formCheckboxProps: IFormCheckboxProps = {
  id: 'formCheckbox',
  name: 'formCheckbox',
  label: 'Form Checkbox',
};

jest.mock('@ionic-native/keyboard', () => ({
  hide: jest.fn(),
}));

jest.mock('../../../../../../automation/PageLocators', () => ({
  locators: {
    jobPageLocators: {
      aFiltersIcon: 'filtersIcon',
    },
    commonLocators: {
      aHeaderText: 'headerText',
    },
    accountSettings: {
      aChangeCountryHeader: 'changeCountryHeader',
    },
  },
}));
describe('Form Checkbox', () => {
  it('snapshot match', () => {
    const {container} = renderWithDyanmicForm(<FormCheckbox {...formCheckboxProps} />, {});
    expect(container).toBeDefined();
    expect(container.firstChild).toMatchSnapshot();

    expect(screen).toMatchSnapshot();
  });

  it('renders in document and a default value and disabled', () => {
    renderWithDyanmicForm(<FormCheckbox {...formCheckboxProps} isDisabled={true} />, {
      defaultValues: {
        formCheckbox: true,
      },
    });
    const checkbox = screen.getByLabelText(formCheckboxProps.label as string);
    expect(screen.getByText('Form Checkbox')).toBeInTheDocument();
    expect((checkbox as HTMLInputElement).disabled).toBeTruthy();
    expect((checkbox as HTMLInputElement).checked).toBeTruthy();
  });

  it('checkbox aria role is applied once (implicitly for input)', () => {
    renderWithDyanmicForm(<FormCheckbox {...formCheckboxProps} />, {});
    expect(screen.getAllByRole('checkbox').length).toEqual(1);
  });

  it('displays error message when errors', async () => {
    const {getByText, dynamicFormMethods} = renderWithDyanmicForm(<FormCheckbox {...formCheckboxProps} />, {
      toPassBack: ['setError'],
    });

    await act(async () => {
      (dynamicFormMethods as UseFormReturn)?.setError(formCheckboxProps.name!, {
        type: 'manual',
        message: 'Checkbox is required',
      });
    });

    expect(screen.getByRole('checkbox')).toBeInTheDocument();
    expect(getByText('Checkbox is required'));
  });
});
