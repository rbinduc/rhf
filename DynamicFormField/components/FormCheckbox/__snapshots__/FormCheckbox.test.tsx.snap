// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Checkbox snapshot match 1`] = `
<div
  class="MuiFormControl-root css-gdldrm-FormControlStyled e1y7dzbg0"
>
  <label
    class="MuiFormControlLabel-root css-tvaf49-FormControlLabelStyled e1y7dzbg2"
    id=""
  >
    <span
      aria-disabled="false"
      class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorPrimary css-5hszho-CheckboxStyled e1y7dzbg3 MuiIconButton-colorPrimary"
    >
      <span
        class="MuiIconButton-label"
      >
        <input
          aria-label="Form Checkbox"
          class="PrivateSwitchBase-input-4"
          data-indeterminate="false"
          id="formCheckbox"
          type="checkbox"
          value=""
        />
        <svg
          fill="none"
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="19"
            rx="4.5"
            stroke="currentColor"
            width="19"
            x="0.5"
            y="0.5"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </span>
    <span
      class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
    >
      Form Checkbox
    </span>
  </label>
</div>
`;

exports[`Form Checkbox snapshot match 2`] = `
Object {
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;
